association.go

```diff
@@ -384,18 +450,118 @@ func (a *Association) handleInbound(raw []byte) error {
 	}
 
 	for _, c := range p.chunks {
-		packets, err := a.handleChunk(p, c)
+		err := a.handleChunk(p, c)
 		if err != nil {
 			return errors.Wrap(err, "failed handling chunk")
 		}
-		for _, p := range packets {
-			err = a.send(p)
-			if err != nil {
-				return errors.Wrap(err, "failed sending reply")
+	}
+
+	return nil
+}
+
+// handleOutbound handles outgoing processes
+func (a *Association) handleOutbound() error {
+	a.lock.Lock()
+
+	packets := []*packet{}
+	if a.controlQueue.size() > 0 {
+		packets = append(packets, a.controlQueue.popAll()...)
+	}
+
+	state := a.getState()
+
+	if state == established {
+		if !a.willRetransmitDataChunks {
+			// Pop unsent data chunks from the pending queue to send as much as
+			// cwnd and rwnd allow.
+			chunks := a.popPendingDataChunksToSend()
+			if len(chunks) > 0 {
+				// Start timer. (noop if already started)
+				a.log.Tracef("[%s] T3-rtx timer start (pt1)", a.name())
+				a.t3RTX.start(a.rtoMgr.getRTO())
+
+				packets = append(packets, a.bundleDataChunksIntoPackets(chunks)...)
+			}
+		} else {
+			a.willRetransmitDataChunks = false
+
+			packets = append(packets, a.getDataPacketsToRetransmit()...)
+		}
+
+		if a.willRetransmitFast {
+			a.willRetransmitFast = false
+
+			toFastRetrans := []chunk{}
+			fastRetransSize := commonHeaderSize
+
+			for i := 0; ; i++ {
+				c, ok := a.inflightQueue.get(a.cumulativeTSNAckPoint + uint32(i) + 1)
+				if !ok {
+					break // end of pending data
+				}
+
+				if c.acked || c.abandoned {
+					continue
+				}
+
+				if c.nSent > 1 || c.missIndicator < 3 {
+					continue
+				}
+
+				// RFC 4960 Sec 7.2.4 Fast Retransmit on Gap Reports
+				//  3)  Determine how many of the earliest (i.e., lowest TSN) DATA chunks
+				//      marked for retransmission will fit into a single packet, subject
+				//      to constraint of the path MTU of the destination transport
+				//      address to which the packet is being sent.  Call this value K.
+				//      Retransmit those K DATA chunks in a single packet.  When a Fast
+				//      Retransmit is being performed, the sender SHOULD ignore the value
+				//      of cwnd and SHOULD NOT delay retransmission for this single
+				//		packet.
+				dataChunkSize := dataChunkHeaderSize + uint32(len(c.userData))
+				if a.mtu-fastRetransSize >= dataChunkSize {
```

Comment:
not sure to understand the spec so not sure if this is ok or not, but here you check if the current chunk fits the mtu.. is that right or you meant something along the lines of `a.mtu-fastRetransSize >= (dataChunkSize + fastRetransSize)`?

Reply:
I guess you meant:
`a.mtu >= (dataChunkSize + fastRetransSize)`
Do you think this is more readable?

Reply:
I'm so stupid, sorry, disregard my comment, I read `a.mtu-fastRetransSize` as a var :man_facepalming: sorry for the noise

Reply:
Done

Reply:
Thanks, @hugo for spending time to review the code. I really appreciate it.